### PR TITLE
Provide `program_id()` syscall

### PR DESCRIPTION
**Pull request by vivekvpandya**
**Release Notes:** `program_id()` function is added to `gstd` which returns program_id of the program in execution. This can be used where the program wants to store funds like fungible-tokens to for itself.

### CODE REVIEW

**Comment by breathx** [comment]
Don't see the use-case of this function.
Anyway it also requires sys_cal in `gcore` and `gstd` libs to be called from programs.

**Comment by vivekvpandya** [comment]
> Don't see the use-case of this function. Anyway it also requires sys_cal in `gcore` and `gstd` libs to be called from programs.

DEX smart contract wants funds of fungible-token in its own account. That is why it needs to know its own program-id.

**Review by breathx — CHANGES_REQUESTED** [review]
New syscall addition always requires demo, showing it's functionality.

Also you need to provide the wrapper around it not only for `gcore`, but for `gstd` also.

It would be good to provide `Release notes` for this changes as well.

Except these things, looks good to me

File: gcore/src/exec.rs [hunk 1]
```diff
@@ -147,3 +149,9 @@ pub fn wake(waker_id: MessageId) {
         sys::gr_wake(waker_id.as_slice().as_ptr());
     }
 }
+
+pub fn actor_id() -> ActorId {
```
Comment (shamilsan) [inline gcore/src/exec.rs:153]: Two concerns:
1. The function name seems to be irrelevant to its purpose. It always returns the program ID so `actor_id` looks too general and confusing.
2. This function needs to be documented.

Comment (vivekvpandya) [reply gcore/src/exec.rs:153]: yes , but I have heard that program_id will be deprecated soon. Also `ActorId` struct is available to be used here, `ProgramId` is not.

Comment (shamilsan) [reply gcore/src/exec.rs:165]: You are right about *types* `ActorId` vs `ProgramId`. But the *function name* should reflect that it does. When I see `exec::actor_id()` in code, it is not obvious to me, who is this actor. So in my opinion `exec::program_id()` or `exec::my_id()` or even `exec::me()` is more speaking for the end user. I'd choose `exec::program_id()` as the most neutral name.

File: gcore/src/exec.rs [hunk 2]
```diff
@@ -20,6 +20,7 @@
 //!
 //! Provides API for low-level async implementation.
```
[the remark this answers is not in the document]
Comment (shamilsan) [reply gcore/src/exec.rs:21]: I can't agree, sorry. `msg::actor_id()` or `msg::program_id()` looks a little bit weird. `exec::program_id()` looks quite logical.

[the remark this answers is not in the document]
Comment (vivekvpandya) [reply gcore/src/exec.rs:21]: @breathx is okay to mark this as resolved now?

**Review by shamilsan — APPROVED** [review]
I don't see the need for this callback itself 😄 but I hope I'll see it later.
